File: pyflow/samples/demo_flow.go
package main

import "fmt"

type Record struct {
	Value int
}

type Pipeline struct {
	Multiplier int
}

func loadRecords() []Record {
	return []Record{{Value: 3}, {Value: -1}, {Value: 7}}
}

func validateRecord(record Record) bool {
	return record.Value >= 0
}

func normalizeRecord(record Record, multiplier int) Record {
	return Record{Value: record.Value * multiplier}
}

func summarizeRecords(records []Record) int {
	total := 0
	for _, record := range records {
		total += record.Value
	}
	return total
}

func saveReport(total int) {
	fmt.Printf("processed total: %d\n", total)
}

func (pipeline Pipeline) process(records []Record) []Record {
	clean := make([]Record, 0, len(records))
	for _, record := range records {
		if validateRecord(record) {
			clean = append(clean, normalizeRecord(record, pipeline.Multiplier))
		}
	}
	return clean
}

func (pipeline Pipeline) run() {
	raw := loadRecords()
	clean := pipeline.process(raw)
	total := summarizeRecords(clean)
	saveReport(total)
}

func say(){
	fmt.Printf("hi")
}

func main() {
	pipeline := Pipeline{Multiplier: 2}
	pipeline.run()
	go say()
	go say()
}

File: pyflow/samples/demo_flow.rs
#[derive(Clone, Copy)]
struct Record {
    value: i32,
}

struct Pipeline {
    multiplier: i32,
}

fn load_records() -> Vec<Record> {
    vec![Record { value: 3 }, Record { value: -1 }, Record { value: 7 }]
}

fn validate_record(record: &Record) -> bool {
    record.value >= 0
}

fn normalize_record(record: Record, multiplier: i32) -> Record {
    Record {
        value: record.value * multiplier,
    }
}

fn summarize_records(records: &[Record]) -> i32 {
    records.iter().map(|record| record.value).sum()
}

fn save_report(total: i32) {
    println!("processed total: {total}");
}

impl Pipeline {
    fn process(&self, records: Vec<Record>) -> Vec<Record> {
        records
            .into_iter()
            .filter(|record| validate_record(record))
            .map(|record| normalize_record(record, self.multiplier))
            .collect()
    }

    fn run(&self) {
        let raw = load_records();
        let clean = self.process(raw);
        let total = summarize_records(&clean);
        save_report(total);
    }
}

fn main() {
    let pipeline = Pipeline { multiplier: 2 };
    pipeline.run();
}
